contracts/lens/Quoter.sol

```diff
@@ -0,0 +1,320 @@
+    function validateRevertReason(bytes memory reason) private pure returns (bytes memory) {
+        if (reason.length < MINIMUM_VALID_RESPONSE_LENGTH) {
+            //if InvalidLockAcquiredSender()
+            if (reason.length == MINIMUM_CUSTOM_ERROR_LENGTH) {
+                assembly {
+                    revert(reason, 4)
+                }
+            }
+            if (reason.length < MINIMUM_REASON_LENGTH) {
+                revert UnexpectedRevertBytes();
+            }
+            assembly {
+                reason := add(reason, 0x04)
+            }
+            revert(abi.decode(reason, (string)));
+        }
+        return reason;
+    }
+
+    /// @dev parse revert bytes from a single-pool quote
+    function _handleRevertSingle(bytes memory reason, PoolKey memory poolKey)
+        private
+        view
+        returns (int128[] memory deltaAmounts, uint160 sqrtPriceX96After, uint32 initializedTicksLoaded)
+    {
+        int24 tickBefore;
+        int24 tickAfter;
+        BalanceDelta deltas;
+        deltaAmounts = new int128[](2);
+        (, tickBefore,) = manager.getSlot0(poolKey.toId());
+        reason = validateRevertReason(reason);
+        (deltas, sqrtPriceX96After, tickAfter) = abi.decode(reason, (BalanceDelta, uint160, int24));
+        deltaAmounts[0] = deltas.amount0();
+        deltaAmounts[1] = deltas.amount1();
+
+        initializedTicksLoaded = PoolTicksCounter.countInitializedTicksLoaded(manager, poolKey, tickBefore, tickAfter);
+    }
+
+    /// @dev parse revert bytes from a potentially multi-hop quote and return the delta amounts, sqrtPriceX96After, and initializedTicksLoaded
+    function _handleRevert(bytes memory reason)
+        private
+        pure
+        returns (
+            int128[] memory deltaAmounts,
+            uint160[] memory sqrtPriceX96AfterList,
+            uint32[] memory initializedTicksLoadedList
+        )
+    {
+        reason = validateRevertReason(reason);
+        (deltaAmounts, sqrtPriceX96AfterList, initializedTicksLoadedList) =
+            abi.decode(reason, (int128[], uint160[], uint32[]));
+    }
+
+    /// @dev quote an ExactInput swap along a path of tokens, then revert with the result
+    function _quoteExactInput(QuoteExactInputParams memory params) public selfOnly returns (bytes memory) {
+        uint256 pathLength = params.path.length;
+
+        int128[] memory deltaAmounts = new int128[](pathLength + 1);
+        uint160[] memory sqrtPriceX96AfterList = new uint160[](pathLength);
+        uint32[] memory initializedTicksLoadedList = new uint32[](pathLength);
+        Currency prevCurrencyOut;
+        uint128 prevAmountOut;
+
+        for (uint256 i = 0; i < pathLength; i++) {
+            (PoolKey memory poolKey, bool zeroForOne) =
+                PathKeyLib.getPoolAndSwapDirection(params.path[i], i == 0 ? params.currencyIn : prevCurrencyOut);
```

Comment:
very up to you but I like the library syntax where you can do 

pathKey.getPoolAndSwapDirection(currencyIn)

and you define above `using PathKeyLib for PathKey`.. but just syntactical

Reply:
idt that's possible because there is no `PathKey` instance here?

Reply:
isn't params.path[i] the PathKey?

Reply:
yea path[i] is PathKey - agreed with sara on this preference